File: lib/ChatScreen.java
import 'package:firebase_auth/firebase_auth.dart';
import 'package:cloud_firestore/cloud_firestore.dart';
import 'package:firebase_storage/firebase_storage.dart';
import 'package:flutter/material.dart';
import 'package:image_picker/image_picker.dart';
import 'dart:io';
import 'auth_service.dart';

class ChatScreen extends StatefulWidget {
  @override
  _ChatScreenState createState() => _ChatScreenState();
}

class _ChatScreenState extends State<ChatScreen> {
  final TextEditingController _controller = TextEditingController();
  final FirebaseAuth _auth = FirebaseAuth.instance;
  final ImagePicker _picker = ImagePicker();
  final AuthService _authService = AuthService();
  late User _currentUser;

  @override
  void initState() {
    super.initState();
    _currentUser = _auth.currentUser!;
  }

  Future<void> sendMessage(String text) async {
    try {
      await FirebaseFirestore.instance.collection('messages').add({
        'text': text,
        'createdAt': FieldValue.serverTimestamp(),
        'userId': _currentUser.uid,
        'userEmail': _currentUser.email,
      });
    } catch (e) {
      print("Error sending message: $e");
    }
  }

  Future<void> uploadFile() async {
    final pickedFile = await _picker.pickImage(source: ImageSource.gallery);
    if (pickedFile != null) {
      File file = File(pickedFile.path);
      try {
        // Upload image to Firebase Storage
        await FirebaseStorage.instance
            .ref('chat_images/${DateTime.now().millisecondsSinceEpoch}.jpg')
            .putFile(file);

        // Get download URL
        String downloadUrl = await FirebaseStorage.instance
            .ref('chat_images/${DateTime.now().millisecondsSinceEpoch}.jpg')
            .getDownloadURL();

        // Save image URL in Firestore as a message
        await FirebaseFirestore.instance.collection('messages').add({
          'imageUrl': downloadUrl,
          'createdAt': FieldValue.serverTimestamp(),
          'userId': _currentUser.uid,
          'userEmail': _currentUser.email,
        });
        print("File uploaded successfully");
      } catch (e) {
        print("Error uploading file: $e");
      }
    }
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text("Firebase Chat"),
        actions: [
          IconButton(
            icon: Icon(Icons.exit_to_app),
            onPressed: () {
              _authService.signOut();
              Navigator.pushReplacementNamed(context, '/login');
            },
          ),
        ],
      ),
      body: Column(
        children: [
          Expanded(
            child: StreamBuilder(
              stream: FirebaseFirestore.instance
                  .collection('messages')
                  .orderBy('createdAt', descending: true)
                  .snapshots(),
              builder: (ctx, AsyncSnapshot<QuerySnapshot> messageSnapshot) {
                if (messageSnapshot.connectionState == ConnectionState.waiting) {
                  return Center(child: CircularProgressIndicator());
                }

                final messageDocs = messageSnapshot.data!.docs;

                return ListView.builder(
                  reverse: true,
                  itemCount: messageDocs.length,
                  itemBuilder: (ctx, index) {
                    final message = messageDocs[index];
                    if (message['text'] != null) {
                      return ListTile(
                        title: Text(message['text']),
                        subtitle: Text(message['userEmail']),
                      );
                    } else if (message['imageUrl'] != null) {
                      return ListTile(
                        title: Image.network(message['imageUrl']),
                        subtitle: Text(message['userEmail']),
                      );
                    }
                    return SizedBox();
                  },
                );
              },
            ),
          ),
          Row(
            children: [
              IconButton(
                icon: Icon(Icons.camera_alt),
                onPressed: uploadFile,
              ),
              Expanded(
                child: TextField(
                  controller: _controller,
                  decoration: InputDecoration(labelText: 'Enter a message'),
                ),
              ),
              IconButton(
                icon: Icon(Icons.send),
                onPressed: () async {
                  sendMessage(_controller.text);
                  _controller.clear();
                },
              ),
            ],
          ),
        ],
      ),
    );
  }
}

File: lib/chat_screen.dart

import 'package:cloud_firestore/cloud_firestore.dart';
import 'package:firebase_auth/firebase_auth.dart';
import 'package:flutter/material.dart';

class ChatScreen extends StatefulWidget {
  final String chatId;
  final String targetUserId;
  final String productId;

  ChatScreen({required this.chatId, required this.targetUserId, required this.productId});

  @override
  _ChatScreenState createState() => _ChatScreenState();
}

class _ChatScreenState extends State<ChatScreen> {
  final TextEditingController messageController = TextEditingController();
  String targetUserEmail = '';
  String targetUsername = '';
  String userId = 'A7FSt1TsLTOGNK1n5u8B9DqNznA82YdW'; //UserID

  // Cache messages locally
  List<QueryDocumentSnapshot> cachedMessages = [];

  @override
  void initState() {
    super.initState();
    _fetchUserId();
    _fetchTargetUserEmail();
  }

  // Fetch current user's ID from Firebase Authentication
  Future<void> _fetchUserId() async {
    final currentUser = FirebaseAuth.instance.currentUser;
    if (currentUser != null) {
      setState(() {
      });
    }
  }

  // Fetch target user's email and username from Firestore
  Future<void> _fetchTargetUserEmail() async {
    try {
      final targetUserDoc = await FirebaseFirestore.instance.collection('users')
          .doc(widget.targetUserId)
          .get();

      if (targetUserDoc.exists) {
        setState(() {
          targetUserEmail = targetUserDoc.data()?['email'] ?? 'No email';
          targetUsername = targetUserDoc.data()?['name'] ?? 'No name';
        });
      } else {
        print("Target user document does not exist.");
      }
    } catch (e) {
      print("Error fetching target user email: $e");
    }
  }

  // Send a message to Firestore
  void sendMessage() async {
    if (messageController.text.trim().isNotEmpty) {
      try {
        await FirebaseFirestore.instance
            .collection('chats')
            .doc(widget.chatId)
            .collection('messages')
            .add({
          'text': messageController.text,
          'senderId': userId,
          'product_id': widget.productId,
          'createdAt': FieldValue.serverTimestamp(),
        });
        print("Message sent: ${messageController.text}"); // Debugging
        messageController.clear();
      } catch (e) {
        print('Error sending message: $e');
      }
    }
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(targetUsername.isNotEmpty ? targetUsername : targetUserEmail),
      ),
      body: Column(
        children: [
          Expanded(
            child: StreamBuilder(
              stream: FirebaseFirestore.instance
                  .collection('chats')
                  .doc(widget.chatId)
                  .collection('messages')
                  .where('product_id', isEqualTo: widget.productId)
              // .orderBy('createdAt', descending: true)
                  .snapshots(),
              builder: (context, AsyncSnapshot<QuerySnapshot> snapshot) {
                if (snapshot.connectionState == ConnectionState.waiting && cachedMessages.isEmpty) {
                  return Center(child: CircularProgressIndicator());
                }

                if (snapshot.hasError) {
                  return Center(child: Text("Error: ${snapshot.error}"));
                }

                if (snapshot.hasData && snapshot.data!.docs.isNotEmpty) {
                  cachedMessages = snapshot.data!.docs;
                  print("Messages fetched: ${snapshot.data!.docs.length}"); // Debugging
                } else {
                  print("No messages found");
                }

                if (cachedMessages.isEmpty) {
                  return Center(child: Text('No messages yet! Start chatting.'));
                }

                return ListView.builder(
                  reverse: true,
                  itemCount: cachedMessages.length,
                  itemBuilder: (context, index) {
                    final message = cachedMessages[index];
                    final isCurrentUser = message['senderId'] == userId;
                    final timestamp = message['createdAt'] != null
                        ? (message['createdAt'] as Timestamp).toDate()
                        : DateTime.now();
                    final formattedTime = '${timestamp.hour}:${timestamp.minute}';

                    return Align(
                      alignment: isCurrentUser ? Alignment.centerRight : Alignment.centerLeft,
                      child: Padding(
                        padding: const EdgeInsets.symmetric(vertical: 4.0, horizontal: 8.0),
                        child: Column(
                          crossAxisAlignment: isCurrentUser
                              ? CrossAxisAlignment.end
                              : CrossAxisAlignment.start,
                          children: [
                            Container(
                              padding: const EdgeInsets.all(10.0),
                              decoration: BoxDecoration(
                                color: isCurrentUser ? Colors.blue : Colors.grey[300],
                                borderRadius: BorderRadius.circular(12),
                              ),
                              child: Text(
                                message['text'] ?? '',
                                style: TextStyle(color: isCurrentUser ? Colors.white : Colors.black),
                              ),
                            ),
                            SizedBox(height: 4),
                            Text(
                              formattedTime,
                              style: TextStyle(fontSize: 10, color: Colors.grey),
                            ),
                          ],
                        ),
                      ),
                    );
                  },
                );
              },
            ),
          ),
          Padding(
            padding: const EdgeInsets.all(8.0),
            child: Row(
              children: [
                Expanded(
                  child: TextField(
                    controller: messageController,
                    decoration: InputDecoration(labelText: 'Send a message...'),
                  ),
                ),
                IconButton(
                  icon: Icon(Icons.send),
                  onPressed: sendMessage,
                ),
              ],
            ),
          ),
        ],
      ),
    );
  }
}


